lib/history/sqlite.go

```diff
@@ -0,0 +1,224 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package history
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strings"
+	"time"
+
+	pb "github.com/gravitational/satellite/agent/proto/agentpb"
+
+	"github.com/gravitational/trace"
+	_ "github.com/mattn/go-sqlite3" // initialize sqlite3
+)
+
+// SQLiteTimeline represents a timeline of cluster status events. The timeline
+// can hold a specified amount of events and uses a FIFO eviction policy.
+// Timeline events are stored in a local sqlite database.
+//
+// Implements Timeline
+type SQLiteTimeline struct {
+	// size specifies the max number of events stored in the timeline.
+	size int
+	// database points to underlying sqlite database.
+	database *sql.DB
+	// lastStatus holds the last recorded cluster status.
+	lastStatus *Cluster
+}
+
+// NewSQLiteTimeline initializes and returns a new SQLiteTimeline with the
+// specified size. Initial cluster status is `Unknown`.
+func NewSQLiteTimeline(database *sql.DB, size int) (Timeline, error) {
+	if err := initEventTable(database); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &SQLiteTimeline{
+		size:       size,
+		database:   database,
+		lastStatus: &Cluster{Status: pb.SystemStatus_Unknown.String()},
```

Comment:
Won't this cause an influx of events on every restart? Maybe the last status should also be stored in the database, and the timeline will initialize itself with it when starting up?

Reply:
I think this might be unavoidable. If the satellite agent restarts I'm not sure it will immediately return to it's previous state before shutting down.
Will do some testing to see how it acts in this situation.